src/app/util/ember-compatibility-functions.cpp

```diff
@@ -94,5 +97,128 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
     return emberAfContainsServer(aEndPointId, aClusterId);
 }
 
+bool ServerClusterAttributeExists(chip::EndpointId aEndPointId, chip::ClusterId aClusterId, chip::AttributeId aAttributeId)
+{
+    // TODO: Currently, we are using cluster catalog from the ember library, this should be modified or replaced after several
+    // updates to Commands.
+    return emberAfReadServerAttribute(aEndPointId, aClusterId, aAttributeId, nullptr, 0) == EMBER_ZCL_STATUS_SUCCESS;
+}
+
+CHIP_ERROR ReadSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVWriter & aWriter)
+{
+    ChipLogDetail(DataManagement,
+                  "Received Cluster Command: Cluster=%" PRIx16 " NodeId=0x" ChipLogFormatX64 " Endpoint=%" PRIx8 " FieldId=%" PRIx8
+                  " ListIndex=%" PRIx8,
+                  aClusterInfo.mClusterId, ChipLogValueX64(aClusterInfo.mNodeId), aClusterInfo.mEndpointId, aClusterInfo.mFieldId,
+                  aClusterInfo.mListIndex);
+    uint8_t data[256];
+    EmberAfAttributeType dataType;
+    EmberAfStatus status = emberAfReadAttribute(aClusterInfo.mEndpointId, aClusterInfo.mClusterId, aClusterInfo.mFieldId,
+                                                CLUSTER_MASK_SERVER, data, sizeof(data), &dataType);
+
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    switch (dataType)
```

Comment:
This switch statement is pretty hard to follow with all of the fall through cases.

Is it possible to simplify this?  I don't think this type of brevity is always useful.  It doesn't meaningfully reduce code space and makes bugs harder to see.

Reply:
Maybe we can introduce some macro / templates here.

I think the code without FALLTHROUGH is more clear, however, the compiler is not happy to see it.

Reply:
Added "BaseType" to make the code clearer here.